packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs

```diff
@@ -45,16 +48,31 @@ where
         transaction: &Transaction,
     ) -> Result<ValidationResult<block_execution_outcome::v0::BlockExecutionOutcome, Error>, Error>
     {
-        let platform_version = platform_state.current_platform_version()?;
+        // TODO: We should do it above and validate version from request
+        //  prepare proposal should set this version
+        // Epoch information is always calculated with the last committed platform version
+        // even if we are switching to a new version in this block.
+        let last_committed_platform_version = platform_state.current_platform_version()?;
 
         let epoch_info = self.gather_epoch_info(
             &block_proposal,
             transaction,
             platform_state,
-            platform_version,
+            last_committed_platform_version,
         )?;
 
-        match platform_version
+        // Determine a protocol version for this block
+        let block_platform_version = if epoch_info.is_epoch_change_but_not_genesis() {
+            // Switch to next proposed platform version if we are on the first block of the new epoch
+            // This version must be set to the state as current one during block processing
+            let protocol_version = platform_state.next_epoch_protocol_version();
+            PlatformVersion::get(protocol_version)?
```

Comment:
This is where things will break if node didn't update.

Reply:
Added panic